src/track/beats.cpp

```diff
@@ -1,29 +1,538 @@
+
+FramePos Beats::getLastBeatPosition() const {
+    return m_beatsInternal.getLastBeatPosition();
+}
+
+SINT Beats::getSampleRate() const {
+    return m_track->getSampleRate();
+}
+
+void Beats::setAsDownbeat(int beatIndex) {
+    QMutexLocker locker(&m_mutex);
+    m_beatsInternal.setAsDownbeat(beatIndex);
+    locker.unlock();
+    emit updated();
+}
+
+QDebug operator<<(QDebug dbg, const BeatsPointer& arg) {
+    dbg << arg->m_beatsInternal;
+    return dbg;
+}
+
+QDebug operator<<(QDebug dbg, const BeatsInternal& arg) {
+    QVector<FramePos> beatFramePositions;
+    for (const auto& beat : arg.m_beats) {
+        beatFramePositions.append(beat.getFramePosition());
+    }
+    dbg << "["
+        << "Aggregate BPM:" << arg.m_bpm << "|"
+        << "Number of beats:" << arg.m_beats.size() << "|"
+        << "Beats:" << beatFramePositions << "]";
+    return dbg;
+}
+Beat BeatsInternal::findNthBeat(FramePos frame, int n) const {
+    if (!isValid() || n == 0) {
+        return kInvalidBeat;
+    }
+
+    Beat beat(frame);
+    // it points at the first occurrence of beat or the next largest beat
+    BeatList::const_iterator it =
+            std::lower_bound(m_beats.cbegin(), m_beats.cend(), beat);
+
+    // If the position is within 1/10th of the average beat length,
+    // pretend we are on that beat.
+    const double kFrameEpsilon =
+            kBeatVicinityFactor * getBeatLengthFrames(getBpm(), m_iSampleRate);
+
+    // Back-up by one.
+    if (it != m_beats.begin()) {
+        --it;
+    }
+
+    // Scan forward to find whether we are on a beat.
+    BeatList::const_iterator on_beat = m_beats.cend();
+    BeatList::const_iterator previous_beat = m_beats.cend();
+    BeatList::const_iterator next_beat = m_beats.cend();
+    for (; it != m_beats.end(); ++it) {
+        FrameDiff_t delta = it->getFramePosition() - beat.getFramePosition();
+
+        // We are "on" this beat.
+        if (abs(delta) < kFrameEpsilon) {
+            on_beat = it;
+            break;
+        }
+
+        if (delta < 0) {
+            // If we are not on the beat and delta < 0 then this beat comes
+            // before our current position.
+            previous_beat = it;
+        } else {
+            // If we are past the beat and we aren't on it then this beat comes
+            // after our current position.
+            next_beat = it;
+            // Stop because we have everything we need now.
+            break;
+        }
+    }
+
+    // If we are within epsilon frames of a beat then the immediately next and
+    // previous beats are the beat we are on.
+    if (on_beat != m_beats.end()) {
+        next_beat = on_beat;
+        previous_beat = on_beat;
+    }
+
+    if (n > 0) {
+        for (; next_beat != m_beats.end(); ++next_beat) {
+            if (n == 1) {
+                // Return a sample offset
+                return *next_beat;
+            }
+            --n;
+        }
+    } else if (n < 0 && previous_beat != m_beats.end()) {
```

Comment:
Here we need to extrapolate beats before the beginning of the track based from the first tempo and downbeat markers.

Reply:
Sure. But what is the lower limit?
The `BeatGrid` generated beats on-demand using the `BeatGridIterator` based on what part of the track is on the screen in the waveform. We cannot store an arbitrarily large number of beats just to have beats before the start of the track.

Reply:
I don't think there should be a lower limit. I think this should generate beats on demand at any arbitrary position before the start of the track like the legacy BeatGridIterator.

Reply:
This would mean creating a hybrid of `BeatMap` and `BeatGrid` for the `BeatsIterator` and all functions like `findNthBeat`.